### DISCUSSION TITLE
How to pass custom colormaps with float values?

### DISCUSSION THREAD

**Discussion by neelduttahere** [developmentseed/titiler#496] [Q&A] [answered]
Hi,

I am trying to pass the following colormap for visualizing in NDVI using the stac endpoint.

This is my url format that I am using to request tiles:

`http://<something>/stac/tiles/WebMercatorQuad/12/3052/1728?url={stacURL}&rescale=-1%2C1&colormap=%7B%22-1%22%3A%22%23a50026%22%2C%220%22%3A%22%23d73027%22%2C%22.1%22%3A%22%23f46d43%22%2C%22.2%22%3A%22%23fdae61%22%2C%22.3%22%3A%22%23fee08b%22%2C%22.4%22%3A%22%23ffffbf%22%2C%22.5%22%3A%22%23d9ef8b%22%2C%22.6%22%3A%22%23a6d96a%22%2C%22.7%22%3A%22%2366bd63%22%2C%22.8%22%3A%22%231a9850%22%2C%22.9%22%3A%22%23006837%22%2C%221%22%3A%22%2302542d%22%7D&colormap_type=linear`

In the above url, I have encode the following colormap in json:

`{"-1":"#a50026","0":"#d73027",".1":"#f46d43",".2":"#fdae61",".3":"#fee08b",".4":"#ffffbf",".5":"#d9ef8b",".6":"#a6d96a",".7":"#66bd63",".8":"#1a9850",".9":"#006837","1":"#02542d"}`

### Issue

I am getting internal server 500 response when requesting tiles with the above color map & url format. To be more detailed, this is the response for 500 status code titiler: `{"detail":"invalid literal for int() with base 10: '0.1'"}` 

Is it unable to accept float values as key to each color? Because that is what the error seems like. Would really love some clarity on how do we pass NDVI/other color maps that use a similar scale (-1,1)  using titiler's endpoint.

**Comment by vincentsarago** [comment, marked as answer]
@neelduttahere `Discrete` colormap assign RGBA color to an integer while you are trying to assign it to a float value. What you want to use is `interval` colormap https://cogeotiff.github.io/rio-tiler/colormap/#intervals-colormaps
https://github.com/developmentseed/titiler/blob/master/src/titiler/core/tests/test_factories.py#L136-L148

**Reply by neelduttahere** [reply]
Thank you for replying, I tried passing the colormap value in this format:

`http://<sample>/stac/tiles/WebMercatorQuad/11/1471/922?url=https://planetarycomputer.microsoft.com/api/stac/v1/collections/sentinel-2-l2a/items/S2B_MSIL2A_20220503T050649_R019_T43QHV_20220503T184327&resampling_method=nearest&expression=(B08-B04)/(B08+B04)&rescale=-1,1&colormap=[([-1.0,-0.2],[0, 0, 0,1]),([-0.2,0],[165, 0, 38,1]),([0,0.1],[215, 48, 39,1]),([0.1,0.2],[244, 109, 67,1]),([0.2,0.3],[253, 174, 97,1]),([0.3,0.4],[254, 224, 139,1]),([0.4,0.5],[255, 255, 191,1]),([0.5,0.6],[217, 239, 139,1]),([0.6,0.7],[166, 217, 106,1]),([0.7,0.8],[102, 189, 99,1]),([0.8,0.9],[26, 152, 80,1]),([0.9,1.0],[0, 104, 55,1])]&colormap_type=linear`

Note: characters are passed encoded, and the above it just an example

However, getting the response `{"detail":"Could not parse the colormap value."}`

I am confused as to what is the format or structure in which we should pass the interval colormaps to the tiles enpoint under stac

**Reply by vincentsarago** [reply]
@neelduttahere : 
- you don't need to `rescale` when you use `interval` colormap because you want to use the `intervals` to define the color. When you use the `rescale` the min->max value will be translated to 0->255
- `colormap_type=linear` is not a valid option 
- your colormap needs to be json encoded 
```
colormap=[[[-1.0,-0.2],[0,0,0,1]],[[-0.2,0],[165,0,38,1]],[[0,0.1],[215,48,39,1]],[[0.1,0.2],[244,109,67,1]],[[0.2,0.3],[253,174,97,1]],[[0.3,0.4],[254,224,139,1]],[[0.4,0.5],[255,255,191,1]],[[0.5,0.6],[217,239,139,1]],[[0.6,0.7],[166,217,106,1]],[[0.7,0.8],[102,189,99,1]],[[0.8,0.9],[26,152,80,1]],[[0.9,1.0],[0,104,55,1]]]
```
- alpha value for colormap should be either 0 or 255

**Reply by neelduttahere** [reply]
Great! it works now. Not adding rescale value makes sense. Thank you for the clarification!

**Comment by mrgeorge** [comment]
> * alpha value for colormap should be either 0 or 255

@vincentsarago , I'm curious about intermediate alpha values - it appears that values >= 1 for alpha render at full opacity like 255. Is there something that prevents use of intermediate values for partial transparency, e.g. 128 for half transparent?

**Reply by vincentsarago** [reply]
in theory you should be able to use values between 0 and 255 Yes but because we are using Numpy masked array to store the image data, the value is either True (0) or False (255) thus we can't have partially masked value 🙁 

We just made the MaskedArray switch because in 4 years no-one mentioned that they were using partial transparency 😭

**Reply by vincentsarago** [reply]
FYI: I'm planning to `fix` this on next titiler release

**Reply by mrgeorge** [reply]
What timing 😄 ! Thanks for the response and for looking into this @vincentsarago
